pincer/commands/chat_command_handler.py

```diff
@@ -0,0 +1,435 @@
+    __delete_guild = "/guilds/{command.guild_id}/commands/{command.id}"
+
+    def __init__(self, client: Client):
+        self.client = client
+        self._api_commands: List[AppCommand] = []
+        _log.debug(
+            "%i commands registered.", len(ChatCommandHandler.register.items())
+        )
+
+        self.__prefix = f"applications/{self.client.bot.id}"
+
+    async def get_commands(self) -> List[AppCommand]:
+        """|coro|
+
+        Get a list of app commands from Discord
+
+        Returns
+        -------
+        List[:class:`~pincer.objects.app.command.AppCommand`]
+            List of commands.
+        """
+        # TODO: Update if discord adds bulk get guild commands
+        guild_commands = await gather(*(
+            self.client.http.get(
+                self.__prefix + self.__get_guild.format(
+                    guild_id=guild.id if isinstance(guild, Guild) else guild
+                )
+            ) for guild in self.client.guilds
+        ))
+        return list(
+            map(
+                AppCommand.from_dict,
+                await self.client.http.get(self.__prefix + self.__get)
+                + [cmd for guild in guild_commands for cmd in guild],
+            )
+        )
+
+    async def remove_command(self, cmd: AppCommand):
+        """|coro|
+
+        Remove a specific command
+
+        Parameters
+        ----------
+        cmd : :class:`~pincer.objects.app.command.AppCommand`
+            What command to delete
+        """
+        # TODO: Update if discord adds bulk delete commands
+        if cmd.guild_id:
+            _log.info(
+                "Removing command `%s` with guild id %d from Discord",
+                cmd.name,
+                cmd.guild_id,
+            )
+        else:
+            _log.info("Removing global command `%s` from Discord", cmd.name)
+
+        remove_endpoint = self.__delete_guild if cmd.guild_id else self.__delete
+
+        await self.client.http.delete(
+            self.__prefix + remove_endpoint.format(command=cmd)
+        )
+
+    async def add_command(self, cmd: AppCommand):
+        """|coro|
+
+        Add an app command
+
+        Parameters
+        ----------
+        cmd : :class:`~pincer.objects.app.command.AppCommand`
+            Command to add
+        """
+        _log.info("Updated or registered command `%s` to Discord", cmd.name)
+
+        add_endpoint = self.__add
+
+        if cmd.guild_id:
+            add_endpoint = self.__add_guild.format(command=cmd)
+
+        await self.client.http.post(
+            self.__prefix + add_endpoint, data=cmd.to_dict()
+        )
+
+    async def add_commands(self, commands: List[AppCommand]):
+        """|coro|
+
+        Add a list of app commands
+
+        Parameters
+        ----------
+        commands : List[:class:`~pincer.objects.app.command.AppCommand`]
+            List of command objects to add
+        """
+        await gather(*map(self.add_command, commands))
```

Comment:
We really should have a gatherstarmap xD

Reply:
YES. I'm gonna make that feature request for asyncio so we can use it in 5 years...